cpp/src/arrow/engine/substrait/plan_internal.cc

```diff
@@ -133,5 +136,18 @@ Result<ExtensionSet> GetExtensionSetFromPlan(const substrait::Plan& plan,
                             registry);
 }
 
+Result<std::unique_ptr<substrait::Plan>> PlanToProto(
```

Comment:
The issue is that there is the same function signature in the `relation_internal.h`. What we do there is output the corresponding serialized `substrait::Rel` for a `Declaration`. Since the Substrait relation model entities are final classes which doesn't extend from a generic relation interface, what we do is create a `substrait::Rel` and fill the corresponding part. This is considered as a partial plan. For instance if the passed `Declaration` is a `scan` we populate the `read` in the `substrait::Rel` object. Then in the `plan_internal.cc` we extract that `read` component and bind to `substrait::Rel` which is considered as the full plan. In the `plan_internal.cc` what we do is we pass the `sink` to the `PlanToProto` to get the `substrait::Rel` which is recursively called until the whole plan is serialized. 

So that's the reason for having this function signature to make clear and avoid compiler errors. Wanted to expose both interfaces to the user so that it can be used accordingly.

Reply:
That makes sense.  So the root problem is that a relation and a plan are both represented in Acero by `compute::Declaration` and so there is ambiguity.  I think the name `PlanToProto` is fine.